src/utils/balanceChanges.ts

```diff
@@ -0,0 +1,183 @@
+import _ from 'lodash'
+
+import { Amount, IssuedCurrencyAmount } from '../models/common'
+import TransactionMetadata, { Node } from '../models/transactions/metadata'
+
+import { dropsToXrp } from './xrpConversion'
+
+interface Balance {
+  currency: string
+  issuer?: string
+  value: string
+}
+
+interface BalanceChange {
+  account: string
+  balance: Balance
+}
+
+interface BalanceChanges {
+  account: string
+  balances: Balance[]
+}
+
+interface Fields {
+  Account?: string
+  Balance?: Amount
+  LowLimit?: IssuedCurrencyAmount
+  HighLimit?: IssuedCurrencyAmount
+  // eslint-disable-next-line @typescript-eslint/member-ordering -- okay here, just some of the fields are typed to make it easier
+  [field: string]: unknown
+}
+
+interface NormalizedNode {
+  // 'CreatedNode' | 'ModifiedNode' | 'DeletedNode'
+  NodeType: string
+  LedgerEntryType: string
+  LedgerIndex: string
+  NewFields?: Fields
+  FinalFields?: Fields
+  PreviousFields?: Fields
+  PreviousTxnID?: string
+  PreviouTxnLgrSeq?: number
+}
+
+function normalizeNode(affectedNode: Node): NormalizedNode {
+  const diffType = Object.keys(affectedNode)[0]
+  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- not quite right, but close enough
+  const node = affectedNode[diffType] as NormalizedNode
+  return {
+    ...node,
+    NodeType: diffType,
+    LedgerEntryType: node.LedgerEntryType,
+    LedgerIndex: node.LedgerIndex,
+    NewFields: node.NewFields,
+    FinalFields: node.FinalFields,
+    PreviousFields: node.PreviousFields,
+  }
+}
+
+function normalizeNodes(metadata: TransactionMetadata): NormalizedNode[] {
+  if (metadata.AffectedNodes.length === 0) {
+    return []
+  }
+  return metadata.AffectedNodes.map(normalizeNode)
+}
+
+function groupByAccount(balanceChanges: BalanceChange[]): BalanceChanges[] {
+  const grouped = _.groupBy(balanceChanges, (node) => node.account)
+  return Object.entries(grouped).map(([account, items]) => {
+    return { account, balances: items.map((item) => item.balance) }
+  })
+}
+
+function getValue(balance: Amount): BigNumber {
+  if (typeof balance === 'string') {
+    return new BigNumber(balance)
+  }
+  return new BigNumber(balance.value)
+}
+
+function computeBalanceChange(node: NormalizedNode): BigNumber | null {
+  let value: BigNumber | null = null
+  if (node.NewFields?.Balance) {
+    value = getValue(node.NewFields.Balance)
+  } else if (node.PreviousFields?.Balance && node.FinalFields?.Balance) {
+    value = getValue(node.FinalFields.Balance).minus(
+      getValue(node.PreviousFields.Balance),
+    )
+  }
+  if (value === null || value.isZero()) {
+    return null
+  }
+  return value
+}
+
+function getXRPQuantity(
+  node: NormalizedNode,
+): { account: string; balance: Balance } | null {
+  const value = computeBalanceChange(node)
+
+  if (value === null) {
+    return null
+  }
+
+  return {
+    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- okay here
```

Comment:
Couldn't this still be undefined, though, if both `Accounts` are undefined?

Reply:
My understanding of these meta fields is that one of them has to be defined. I might be wrong though, can @mDuo13 confirm?